app/todos/upsert/upsert-todo.controller.js

```diff
@@ -3,7 +3,119 @@
 angular.module('todoApp.todos')
     .controller('UpsertTodoController', [
         '$scope',
-        function($scope) {
+        '$state',
+        '$stateParams',
+        '$http',
+        '$q',
+        function ($scope,
+                  $state,
+                  $stateParams,
+                  $http,
+                  $q) {
+
+            var vm = this;
+            vm.tagsLoading = true;
+            vm.loading = false;
+            vm.checkboxesTouched = false;
+            vm.isNew = ($stateParams.id === '');
+            vm.todo = null;
+            vm.tags = [];
+
+
+            vm.someTagsSelected = someTagsSelected;
+            vm.upsertTodo = upsertTodo;
+
+            init();
+
+            function init() {
+                vm.tagsLoading = true;
+                $q.all([getTodo(), getTags()])
+                    .then(function (responses) {
+                        vm.todo = responses[0].data;
+                        vm.tags = mapTags(responses[1].data.tags);
+                        console.log(vm.todo);
+                        vm.tagsLoading = false;
+                    })
+                    .catch(function (error) {
+                        console.log('Error loading tags/todo');
+                        vm.tagsLoading = false;
+                    });
+            }
+
+            function getTags() {
+                return $http.get('http://localhost:3000/tags')
+            }
+
+            function getTodo() {
+                var defer = $q.defer();
+
+                if (vm.isNew) {
+                    defer.resolve({data: {title: '', tag_ids: []}});
```

Comment:
In case it is an **Add Todo** form, an empty todo object having empty properties is created.